### PR TITLE
[BLOG] Towards Optimized Microservices Performance &  Sustainability via Istio, Kepler and Smart Scheduling 

### PR DESCRIPTION
**Pull request by jiangphcn**
The related issue is 

https://github.com/cncf/tag-env-sustainability/issues/241

@rootfs

### CODE REVIEW

File: website/content/blog/Towards-Optimized-Microservices-Performance-Sustainability-via-Istio-Kepler-and-Smart-Scheduling.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# Towards Optimized Microservices Performance &  Sustainability via Istio, Kepler and Smart Scheduling
```
Comment (leonardpahlke) [inline website/content/blog/Towards-Optimized-Microservices-Performance-Sustainability-via-Istio-Kepler-and-Smart-Scheduling.md:1]:
```suggestion
---
 title: Towards Optimized Microservices Performance & Sustainability via Istio, Kepler and Smart Scheduling
 description: Optimized Microservices Performance & Sustainability via Istio, Kepler and Smart Scheduling
 slug: 2023-sustainability-istio-kepler-smart-scheduling
 weight: 1
 date:   2023-10-12 00:00:00 +0000
 author: Peng Hui Jiang
 ---
 
 *Guest post as part of the [cloud native sustainability week](https://tag-env-sustainability.cncf.io/cloud-native-sustainability-week/)*
```

File: website/content/blog/Towards-Optimized-Microservices-Performance-Sustainability-via-Istio-Kepler-and-Smart-Scheduling.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+
+## Kepler and Power Modeling
+
+Kepler, known as the "Kubernetes-based Efficient Power Level Exporter," operates by utilizing eBPF technology to examine CPU performance counters and Linux kernel tracepoints. These gathered data, including information from BPF context switch events and sysfs, are inputted into machine learning models. This process enables us to estimate the power consumption of Kubernetes Pods accurately. From its inception, Kepler adheres to three fundamental principles: it is designed to be ubiquitous, capable of running on various platforms like bare-metal or virtual machines, supporting different architectures such as x86, ARM, or S390. Additionally, it is lightweight, ensuring a small footprint and low overhead, and it is grounded in scientific research, relying on well-studied principles.
+
+![Kepler Architecture](images/241-kepler.png)
```
Comment (leonardpahlke) [inline website/content/blog/Towards-Optimized-Microservices-Performance-Sustainability-via-Istio-Kepler-and-Smart-Scheduling.md:28]:
```suggestion
<p class="mt-5 mb-5"><img src="/images/blogs/2023-09-cloud-native-sustainability-week/241-kepler.png"></p>
```

@jiangphcn could you move the image to `/images/blogs/2023-09-cloud-native-sustainability-week/241-kepler.png`

File: website/content/blog/Towards-Optimized-Microservices-Performance-Sustainability-via-Istio-Kepler-and-Smart-Scheduling.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+Scenario 1: Default Scheduling Policy
+
+In the first scenario, we employ Kubernetes' default scheduling policy, allowing the platform's innate intelligence to allocate resources. The "Business Logic" Service and the "Data Store" Service are distributed across Worker01 and Worker02 based on Kubernetes' default algorithm. This scenario relies on Kubernetes' understanding of workload demands, node capacities, and resource availability. The approach embodies simplicity, leveraging Kubernetes' built-in logic to ensure a balanced distribution of services. However, while default scheduling is convenient, it may not always cater to specific workload nuances, potentially leading to suboptimal resource utilization and performance imbalances.
+
+Scenario 2: Customized Service Allocation
+
+In contrast, the second scenario adopts a more targeted approach. Here, the "Data Store" Service is strategically placed on Worker01, while the "Business Logic" Service finds its home on Worker02. This customization allows for meticulous resource allocation, aligning services with nodes that best suit their computational and memory needs.
+
+![Typical Scheduling Policy](images/241-scheduling.png)
```
Comment (leonardpahlke) [inline website/content/blog/Towards-Optimized-Microservices-Performance-Sustainability-via-Istio-Kepler-and-Smart-Scheduling.md:48]:
```suggestion
<p class="mt-5 mb-5"><img src="/images/blogs/2023-09-cloud-native-sustainability-week/241-scheduling.png"></p>
```

@jiangphcn could you move the image to `/images/blogs/2023-09-cloud-native-sustainability-week/241-scheduling.png`

File: website/content/blog/Towards-Optimized-Microservices-Performance-Sustainability-via-Istio-Kepler-and-Smart-Scheduling.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+In the Kubernetes cluster environment, Smart Scheduling executes by scheduling pods onto nodes as recommended by its policies. To assess its performance, an evaluation method is employed, generating corresponding rewards. The process involves key steps in reinforcement learning. The state, denoted as S, signifies the agent's current position within the Kubernetes cluster. Actions (A) represent decisions made by Smart Scheduling, including pod scheduling and scaling. Each action yields a reward, determined by sustainability, performance, and other factors, weighted accordingly (`c_sus`, `c_perf`, `c_res`, `w_sus`, `w_perf`, `w_res`). Episodes conclude when the agent reaches a terminating state, unable to take further actions.
+
+Temporal Difference, a specific formula, calculates the Q-Value, gauging the effectiveness of an action (A) taken at a given state (S). This value, denoted as `Q(A, S)`, is iteratively updated using the Bellman equation and temporal differences. The objective is to minimize state transfer costs, optimizing Smart Scheduling’s decisions within the Kubernetes cluster.
+
+![Typical Scheduling Policy](images/241-test-result.png)
```
Comment (leonardpahlke) [inline website/content/blog/Towards-Optimized-Microservices-Performance-Sustainability-via-Istio-Kepler-and-Smart-Scheduling.md:58]:
```suggestion
<p class="mt-5 mb-5"><img src="/images/blogs/2023-09-cloud-native-sustainability-week/241-test-result.png"></p>
```

@jiangphcn could you move the image to `/images/blogs/2023-09-cloud-native-sustainability-week/241-test-result.png`

**Review by leonardpahlke — COMMENTED** [review]
Thanks for opening the PR @jiangphcn 👍 🙌 

format review comments:

**Comment by leonardpahlke** [comment]
cc @catblade

File: website/content/blog/Towards-Optimized-Microservices-Performance-Sustainability-via-Istio-Kepler-and-Smart-Scheduling.md (hunk 1 above)
Comment (leonardpahlke) [reply website/content/blog/Towards-Optimized-Microservices-Performance-Sustainability-via-Istio-Kepler-and-Smart-Scheduling.md:1]: @jiangphcn feel free to update the description with whatever you feel works best here, thanks!

**Comment by jiangphcn** [comment]
Thanks @leonardpahlke for review and giving valuable comments on this PR. I already addressed them and pushed new commit [4cb6f55](https://github.com/cncf/tag-env-sustainability/pull/242/commits/4cb6f552c6aa15991f783f7b8f29c15bc5bf1f37). 

In the meanwhile, I addressed the markdownlint issue in this PR. Just FYI: there are still some issues in `reducing-your-environmental-impact-with-the-linkerd-service-mesh.md` blog.

**Comment by leonardpahlke** [comment]
@jiangphcn thanks - I we don't need to fix these Lint findings in this PR; I will open another one for that, thanks for pointing it out.

cc @catblade (if you find time for a blog review)

**Comment by catblade** [comment]
Will do tomorrow.  Thanks for the ping!

File: website/content/blog/Towards-Optimized-Microservices-Performance-Sustainability-via-Istio-Kepler-and-Smart-Scheduling.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+## Kepler and Power Modeling
+
+Kepler, known as the "Kubernetes-based Efficient Power Level Exporter," operates by utilizing eBPF technology to examine CPU performance counters and Linux kernel tracepoints. These gathered data, including information from BPF context switch events and sysfs, are inputted into machine learning models. This process enables us to estimate the power consumption of Kubernetes Pods accurately. From its inception, Kepler adheres to three fundamental principles: it is designed to be ubiquitous, capable of running on various platforms like bare-metal or virtual machines, supporting different architectures such as x86, ARM, or S390. Additionally, it is lightweight, ensuring a small footprint and low overhead, and it is grounded in scientific research, relying on well-studied principles.
+
+![Kepler Architecture](images/blogs/2023-09-cloud-native-sustainability-week/241-kepler.png)
```
Comment (guidemetothemoon) [inline website/content/blog/Towards-Optimized-Microservices-Performance-Sustainability-via-Istio-Kepler-and-Smart-Scheduling.md:36]:
```suggestion
<p class="mt-5 mb-5"><img src="/images/blogs/2023-09-cloud-native-sustainability-week/241-kepler.png" alt="Kepler Architecture"></p>
```

File: website/content/blog/Towards-Optimized-Microservices-Performance-Sustainability-via-Istio-Kepler-and-Smart-Scheduling.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+In the first scenario, we employ Kubernetes' default scheduling policy, allowing the platform's innate intelligence to allocate resources. The "Business Logic" Service and the "Data Store" Service are distributed across Worker01 and Worker02 based on Kubernetes' default algorithm. This scenario relies on Kubernetes' understanding of workload demands, node capacities, and resource availability. The approach embodies simplicity, leveraging Kubernetes' built-in logic to ensure a balanced distribution of services. However, while default scheduling is convenient, it may not always cater to specific workload nuances, potentially leading to suboptimal resource utilization and performance imbalances.
+
+Scenario 2: Customized Service Allocation
+
+In contrast, the second scenario adopts a more targeted approach. Here, the "Data Store" Service is strategically placed on Worker01, while the "Business Logic" Service finds its home on Worker02. This customization allows for meticulous resource allocation, aligning services with nodes that best suit their computational and memory needs.
+
+![Typical Scheduling Policy](images/blogs/2023-09-cloud-native-sustainability-week/241-scheduling.png)
```
Comment (guidemetothemoon) [inline website/content/blog/Towards-Optimized-Microservices-Performance-Sustainability-via-Istio-Kepler-and-Smart-Scheduling.md:56]:
```suggestion
<p class="mt-5 mb-5"><img src="/images/blogs/2023-09-cloud-native-sustainability-week/241-scheduling.png" alt="Typical Scheduling Policy"></p>
```

File: website/content/blog/Towards-Optimized-Microservices-Performance-Sustainability-via-Istio-Kepler-and-Smart-Scheduling.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+Temporal Difference, a specific formula, calculates the Q-Value, gauging the effectiveness of an action (A) taken at a given state (S). This value, denoted as `Q(A, S)`, is iteratively updated using the Bellman equation and temporal differences. The objective is to minimize state transfer costs, optimizing Smart Scheduling’s decisions within the Kubernetes cluster.
+
+![Typical Scheduling Policy](images/blogs/2023-09-cloud-native-sustainability-week/241-scheduling.png)
```
Comment (guidemetothemoon) [inline website/content/blog/Towards-Optimized-Microservices-Performance-Sustainability-via-Istio-Kepler-and-Smart-Scheduling.md:66]:
```suggestion
<p class="mt-5 mb-5"><img src="/images/blogs/2023-09-cloud-native-sustainability-week/241-test-result.png" alt="Scheduling Policy Result"></p>

```

**Review by guidemetothemoon — COMMENTED** [review]
@jiangphcn thanks a lot for your contribution!
I have added a few suggestions, mostly related to image references☺️
Markdown errors related to the Linkerd blog post have been resolved in a different PR so the build state should look much better once you sync your branch with main branch.

**Comment by jiangphcn** [comment]
@guidemetothemoon, thanks a lot for your comments and suggestion. I already addressed them in latest commit. Also i fetched the latest from upstream, and don't see markdown lint issue any more.

**Review by leonardpahlke — APPROVED** [review]
LGTM

**Review by guidemetothemoon — APPROVED** [review]
LGTM, thanks!☺️

**Comment by jiangphcn** [comment]
@leonardpahlke @guidemetothemoon thanks for review and merge. I checked the published blog, and it looks that images are not displayed correctly. Could you please help?

**Comment by jiangphcn** [comment]
https://tag-env-sustainability.cncf.io/blog/2023-sustainability-istio-kepler-smart-scheduling/

**Comment by guidemetothemoon** [comment]
@jiangphcn images were placed in the wrong directory, they needed to be placed in the static/images folder. The issue will be fixed once #250 is merged.

**Comment by jiangphcn** [comment]
thank you @guidemetothemoon

**Comment by leonardpahlke** [comment]
Fix Merged

**Pull request merged**